### PR TITLE
Add edgecontext package

### PR DESCRIPTION
**Pull request by fishy**
https://github.com/reddit/baseplate.go/issues/33

Note that event fields are not implemented yet. They depend on the experiment implementation.

### CODE REVIEW

File: edgecontext/edgecontext.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+func FromThriftContext(ctx context.Context) (*EdgeRequestContext, error) {
+	header, ok := thrift.GetHeader(ctx, thriftbp.HeaderEdgeRequest)
+	if !ok {
+		return nil, ErrNoHeader
+	}
+
+	request := baseplate.NewRequest()
+	deserializer := deserializerPool.Get().(*thrift.TDeserializer)
+	defer func() {
+		// See also: https://github.com/apache/thrift/pull/1987
```
Comment (fishy) [inline edgecontext/edgecontext.go]: This is merged in upstream, but I'll do the cleanup in a follow up PR instead, as we have other places that also need to migrate to the new `TSerializerPool`.

File: edgecontext/timestamp.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+const millisecondsPerSecond = int64(time.Second / time.Millisecond)
+
+// TimestampMillisecond implements json encoding/decoding using milliseconds
+// since EPOCH.
+type TimestampMillisecond time.Time
+
+var (
+	_ json.Unmarshaler = (*TimestampMillisecond)(nil)
+	_ json.Marshaler   = TimestampMillisecond{}
+)
```
Comment (pacejackson) [inline edgecontext/timestamp.go]: 🔕 Should we have a `lib` or `utils` package for more generic things like this?  There's not really a reason for it to be in `edgecontext`.

File: edgecontext/edgecontext.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+	"github.com/apache/thrift/lib/go/thrift"
+)
+
+// ErrLoIDWrongPrefix is an error could be returned by New() when passed in LoID
+// does not have the correct prefix.
+var ErrLoIDWrongPrefix = errors.New("edgecontext: loid should have t2_ prefix")
+
+// ErrNoHeader is an error could be returned by FromThriftContext() when passed
+// in context does not have Edge-Request header set.
+var ErrNoHeader = errors.New("edgecontext: no Edge-Request header found")
```
[the remark this answers is not in the document]
Comment (fishy) [reply edgecontext/edgecontext.go]: I usually only have an `errors.go` for error types. For variables that's an overkill.

File: edgecontext/timestamp.go (hunk 2 above)
Comment (fishy) [reply edgecontext/timestamp.go]: Agreed, but I'll do that in a followup PR instead. We also have the microsecond types used in tracing that should be moved out as well.

Comment (konradreiche) [reply edgecontext/timestamp.go]: @pacejackson Not a big fan of generic packages, they become bit a dumping ground

Comment (fishy) [reply edgecontext/timestamp.go]: The package certainly shouldn't be named something too generic like `lib` or `utils`. I'm thinking about `timebp`.

Comment (konradreiche) [reply edgecontext/timestamp.go]: Yes. By the way, it seems the suffixing of `bp` is spreading across the repository. I find it makes the folder structure a lot harder to read. Wouldn't it be possible to just name those things what they are, i.e. `time` and if there's an import conflict, let the user rename it if necessary.

Comment (fishy) [reply edgecontext/timestamp.go]: most of those are the cases that importing conflict is bound to happen. for example metricsbp itself imported go-kit's metrics package, and timebp will certainly import stdlibs time library. I actually like the -bp suffix :)

**Pull request merged**